Implement Reset Hits feature to Weblinks

mahmoudmagdy1-1 opened the pull request:
Pull Request for Issue #475 .

### Summary of Changes

This pull request adds a "Reset Hits" feature to the Joomla Weblinks admin interface when editing a weblink. It adds a new button in the `Publishing` tab to reset a weblink’s hit counter to zero.

[Screencast from 03-25-2025 03:14:31 PM.webm](https://github.com/user-attachments/assets/be34623c-f09c-4853-93bc-8dc3df8d3b12)

### Testing Instructions

1. Login as Administrator
2. Turn on counter clicks from the global configuration for weblinks extension
3. Create a test weblink as an admin
4. Increment the counter from visiting the links in the frontend
5. Edit the test weblink as an admin and hit the "Reset Hits"

### Expected result
Clicking "Reset Hits" when editing a weblink in the 'Publishing' tab sets the weblink’s hit count to 0.

### Actual result
There was no option to reset the hits counter.


### Documentation Changes Required

alikon commented:
@mahmoudmagdy1-1 
please move this pr on the new branch 5.x-dev cause is a new feature

thanks in advance

mahmoudmagdy1-1 commented:
> i don't like where the button reset hit it is placed, it should be on the toolbar

![Screenshot from 2025-03-26 18-56-49](https://github.com/user-attachments/assets/7b50c00c-4215-4a52-bea5-499d8b37ce4a)
you mean right here maybe?

alikon commented:
yes exaclty, cause i think it's much more "family feeling"

mahmoudmagdy1-1 commented:
The latest changes I've made changes the button position on the front end and it handles the cases that if the weblinks user have set their `Count Clicks` in the Global Configuation for Weblinks to Off or if the user is making a new weblink, they wouldn't see the button, this is the final result on the front end
![image](https://github.com/user-attachments/assets/03550719-9c78-4a7b-98df-d578cf3295ca)
The build kept failing so I fixed it and force pushed the commits back to keep the commit history clean, is there an issue with that or that is fine?

mahmoudmagdy1-1 commented:
> so we should not show the reset button if is set to NO in the weblink item options, like is done for the Global Configuration
> 
> and in the other way around

It should be able to handle this case now

joomdonation commented on src/administrator/components/com_weblinks/src/Controller/WeblinkController.php:
```diff
@@ -127,4 +128,33 @@ protected function postSaveHook(BaseDatabaseModel $model, $validData = [])
+      */
+    public function resetHit()
+    {
+        $this->model = $this->getModel();
```
I don't think you need to assign the model to model property of the class here. Just store it in $model variable is enough.

joomdonation commented on src/administrator/components/com_weblinks/src/Controller/WeblinkController.php:
```diff
@@ -127,4 +128,33 @@ protected function postSaveHook(BaseDatabaseModel $model, $validData = [])
+        $this->model = $this->getModel();
+        $id          = $this->input->getInt('id');
+
+        $data = $this->getModel()->getItem($id)->getProperties();
```
Few things here:

- You do not need to call $this->getModel() here again to get model object. Just use existing $model variable above
- Do not call ->getProperties here. Joomla is removing the usage of CMSObject on getItem method, so the code won't work on J6, use get_object_vars function for the time being. Also, we should check and make sure getItem return actual record instead and throw error if the item not found.

joomdonation commented on src/administrator/components/com_weblinks/src/Controller/WeblinkController.php:
```diff
@@ -127,4 +128,33 @@ protected function postSaveHook(BaseDatabaseModel $model, $validData = [])
+            throw new \Exception(Text::_('JERROR_ALERTNOAUTHOR'), 403);
+        }
+
+        if ($this->model->resetHit($id)) {
```
Similar with above, use local variable $model instead of $this->model

joomdonation commented on src/administrator/components/com_weblinks/src/Model/WeblinkModel.php:
```diff
@@ -409,4 +409,26 @@ private function canCreateCategory()
+        }
+
+        $table = $this->getTable();
+        $table->load($pk);
```
We should check and in case $table->load return false, we should throw error here (maybe 404 exception)

joomdonation reviewed (commented):
Thanks for working on this. There are some changes needed. See below:

mahmoudmagdy1-1 commented:
Thanks for the feedback I really appreciate it, I just addressed the issues you mentioned, 
Should I add comments inside those functions? or just stick with the docblock only?

joomdonation commented:
Also, you made too many unnecessary docblock changes in the PR. Please revert all the changes you made to docblock in this PR. For example, for @param spaces, we use the following rules:

- Between tag and type: 3 spaces
- Between type and name: 2 spaces
- Between name and description: 2 space

mahmoudmagdy1-1 commented:
> me and @joomdonation have had a discussion about this pull request, and we come to the conclusion that it would better, to put the reset hit button like it has been done for the banners component

I didn't know about this, thanks for mentioning it, now after the latest changes I changed the implementation to match the banners component's reset

> Also, you made too many unnecessary docblock changes in the PR. Please revert all the changes you made to docblock in this PR. For example, for @param spaces, we use the following rules:

Sorry, I just reverted those changes, thanks for letting me know.

alikon commented:
i'm afraid you have removed too much needed files, please check

mahmoudmagdy1-1 commented:
> i'm afraid you have removed too much needed files, please check

[Screencast from 04-07-2025 09:54:15 AM.webm](https://github.com/user-attachments/assets/eff6049b-9600-4421-90dc-e8b274214659)

It works for me, can you tell me what exactly is not working for you?

joomdonation commented:
Guess it is because you have old modified files left in your installation. Please check the modified files for your PR https://github.com/joomla-extensions/weblinks/pull/590/files , you removed the changes you made in model, controller classes, so the change was lost and won't work anymore.

alikon commented:
i can confirm it works fine 
for those want to test the pr with com_patchtester  this pr https://github.com/joomla-extensions/patchtester/pull/376 must be applied before

mahmoudmagdy1-1 commented:
@joomdonation please take a look when you have time

alikon commented on src/administrator/components/com_weblinks/src/Field/HitsField.php:
```diff
@@ -0,0 +1,52 @@
+ * Hits field for Weblinks.
+ * Renders a read-only input for hits and a reset button.
+ *
+ * @since  5.0.0
```
i don't remember if there is a correct placeholder like __DEPLOY_VERSION__ can you check it instead of hardcoded 5.0.0 ?

mahmoudmagdy1-1 replied:
Thanks for the feedback, I checked for the placeholder and added it in the last commit

mahmoudmagdy1-1 commented:
> Thanks @mahmoudmagdy1-1 for your work. I expected to have server side code to reset hits of weblink item
> But maybe it is not that important to prevent that behavior, so we will accept this change for now.

No problem, Yes I've noticed that but that's how is the reset button is implemented in other components, maybe we will fix this behavior in the future,


> Thanks again and sorry for slow getting back to you, I'm quite busy these days. Looking forward to seeing other improvements from you.

It's Okay, Thanks for taking the time to review and giving me feedback :)
